captum/optim/_core/loss.py

```diff
@@ -103,40 +220,42 @@ def __call__(self, targets_to_values: ModuleOutputMapping) -> torch.Tensor:
         return torch.sum(torch.abs(x_diff)) + torch.sum(torch.abs(y_diff))
 
 
-class L1(Loss):
+@loss_wrapper
+class L1(SimpleLoss):
     """
     L1 norm of the target layer, generally used as a penalty.
     """
 
     def __init__(self, target: nn.Module, constant: float = 0.0) -> None:
-        super(Loss, self).__init__()
-        self.target = target
+        SimpleLoss.__init__(self, target)
         self.constant = constant
 
     def __call__(self, targets_to_values: ModuleOutputMapping) -> torch.Tensor:
         activations = targets_to_values[self.target]
         return torch.abs(activations - self.constant).sum()
 
 
-class L2(Loss):
+@loss_wrapper
+class L2(SimpleLoss):
     """
     L2 norm of the target layer, generally used as a penalty.
     """
 
     def __init__(
         self, target: nn.Module, constant: float = 0.0, epsilon: float = 1e-6
     ) -> None:
-        self.target = target
+        SimpleLoss.__init__(self, target)
         self.constant = constant
         self.epsilon = epsilon
 
     def __call__(self, targets_to_values: ModuleOutputMapping) -> torch.Tensor:
         activations = targets_to_values[self.target]
-        activations = (activations - self.constant).sum()
+        activations = ((activations - self.constant) ** 2).sum()
         return torch.sqrt(self.epsilon + activations)
 
 
-class Diversity(Loss):
+@loss_wrapper
+class Diversity(SimpleLoss):

```

Comment:
The diversity objective is broken for some reason and it does not appear to be because of the objective itself.

Reply:
Do we have test cases on this ? Do you have a reference when it is failing ?

Reply:
I was testing with variations of this code:

```
def visualize(model, target, channel):
    image = optimviz.images.NaturalImage((128, 128), batch=4).to(device)
    transforms = torch.nn.Sequential(
        torch.nn.ReflectionPad2d(16),
        optimviz.transform.RandomSpatialJitter(4),
        optimviz.transform.RandomScale(scale=(1, 0.975, 1.025, 0.95, 1.05)),
        torchvision.transforms.RandomRotation(degrees=(-5,5)),
        optimviz.transform.RandomSpatialJitter(2),
        optimviz.transform.CenterCrop((128, 128)),
    )
    loss_fn = optimviz.loss.ChannelActivation(target, channel) - (1e2 * optimviz.loss.Diversity(target))
    obj = optimviz.InputOptimization(model, loss_fn, image, transforms)
    history = obj.optimize(optimviz.objectives.n_steps(128))
    img = torchvision.utils.make_grid(image(), nrow=image().shape[0])
    return img

grid_img = visualize(model, model.mixed5a.conv_1x1, 9)
```


And you can compare the results to the Lucid code that does the same thing:


https://colab.research.google.com/github/tensorflow/lucid/blob/master/notebooks/feature-visualization/neuron_diversity.ipynb

If you want to run the Lucid notebook, make sure that you add `%tensorflow_version 1.x` on the line before the `import tensorflow as tf` line.

Reply:
Thank you!

Reply:
@ProGamerGov is this still a problem? I seem to be getting okay results with

```
ChannelActivation(model.mixed5a.conv_1x1, 9) + 1e2 * Diversity(model.mixed5a.conv_1x1)
```
![Screen Shot 2021-03-25 at 12 27 33 PM](https://user-images.githubusercontent.com/13292973/112418991-86efae00-8d65-11eb-9ce8-a556d9635923.png)

Reply:
So, I make it work with the Places365 model that I've been playing around with:

![diversity_p365](https://user-images.githubusercontent.com/10626398/112570448-ad383c80-8dab-11eb-944c-110739748cd7.png)

But I was unable to make it work with your code and the visualization function that I shared above. So, now I'm curious what I'm doing wrong as I probably won't be the last person to encounter this.

Reply:
My Colab setup code:

```
!git clone https://github.com/greentfrapp/captum 
%cd captum
!git checkout "optim-wip-objectives-images"

!pip3 install -e .
import sys
sys.path.append('/content/captum')
%cd ..

import warnings
warnings.filterwarnings("ignore", category=UserWarning, module='torch')

%load_ext autoreload
%autoreload 2

import torch
import torchvision
import numpy as np
import matplotlib.pyplot as plt

from captum import attr
from captum.optim._models.inception_v1 import googlenet

import captum.optim as optimviz

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
model = googlenet(pretrained=True).to(device).eval()

from typing import Optional, Tuple, Callable, Dict, List

def show(
    x: torch.Tensor, figsize: Optional[Tuple[int, int]] = None, scale: float = 255.0
) -> None:
    """
    Show CHW & NCHW tensors as an image.
    Args:
        x (torch.Tensor): The tensor you want to save as an image.
        figsize (Tuple[int, int], optional): height & width to use
            for displaying the image figure.
        scale (float): Value to multiply input tensor by so that
            it's value range is [0-255] for display.
    """
    assert x.dim() == 3 or x.dim() == 4
    x = x[0] if x.dim() == 4 else x
    x = x.cpu().permute(1, 2, 0) * scale
    plt.box(False)
    plt.axis("off")
    if figsize is not None:
        plt.figure(figsize=figsize)
    plt.imshow(x.detach().numpy().astype(np.uint8))
    plt.show()
```

My visualization code:

```
def visualize(model, target, channel):
    image = optimviz.images.NaturalImage((128, 128), batch=4).to(device)
    transforms = torch.nn.Sequential(
        torch.nn.ReflectionPad2d(16),
        optimviz.transform.RandomSpatialJitter(4),
        optimviz.transform.RandomScale(scale=(1, 0.975, 1.025, 0.95, 1.05)),
        torchvision.transforms.RandomRotation(degrees=(-5,5)),
        optimviz.transform.RandomSpatialJitter(2),
        optimviz.transform.CenterCrop((128, 128)),
    )
    loss_fn = optimviz.loss.ChannelActivation(target, channel) + 1e2 * optimviz.loss.Diversity(target)
    obj = optimviz.InputOptimization(model, loss_fn, image, None)
    history = obj.optimize(optimviz.optimization.n_steps(128))
    img = torchvision.utils.make_grid(image(), nrow=image().shape[0])
    show(img)

visualize(model, model.mixed5a.conv_1x1, 9)
```

Reply:
@greentfrapp maybe the problem with my code will be more obvious to you?

Reply:
The same code also didn't let me reproduce the negative neuron results unless I did each of the two visualizations individually.

Reply:
@NarineK 

Each branch is concatenated at the end, and Lucid concatenates the branches before the ReLU layer whereas version of the model has the ReLU layers before the concatenation:

```
c1x1 = F.relu(c1x1)
c3x3 = F.relu(c3x3)
c5x5 = F.relu(c5x5)
px = F.relu(px)
return torch.cat([c1x1, c3x3, c5x5, px], dim=1)
```

Reply:
@NarineK @greentfrapp I think I know the cause of the issue, Earlier today I was playing around with a neuron in the Places365 model that seemed to be extremely polysemantic. As I was changing the batch size and rerunning things, it sometimes ended up with all the images looking the same. This was the same error that I had originally, and thus it seems like it may have something to do with the random starting values of the image parameterization.

I believe that it could be caused by a design decision I made earlier on. For each random parameterization, Lucid has each random image initialization in the batch created separately whereas I made Captum copy a single random initialization to each batch value.

Reply:
I think that we can probably change it so that each image in the batch has different random initialization values. The existing alignment objective and SharedImage parameterization wrapper are already capable of ensuring the the similarity we need, and thus we can change the image parameterizations.

Reply:
@NarineK I'm not sure that using diversity on a target before or after the ReLU operation would have a major difference? Though that may be dependent on the target itself.

Reply:
@ProGamerGov I've updated FFTImage and PixelImage to use independent random initializations for each image in a batch. Let me know if this helps?

Reply:
Update on this:

@ProGamerGov can I confirm this is the bug that you see when you mention that the Diversity and negated losses do not work?
![image](https://user-images.githubusercontent.com/13292973/112712964-9dd10480-8f0d-11eb-8952-759c5e50b436.png)

If that's the case, I think the cause is not the random initialization. It is because of what you mentioned earlier re. concating before versus after the ReLU.

This is what I get after commenting out the relu functions in `InceptionModule` i.e.

```
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        c1x1 = self.conv_1x1(x)
        # c1x1 = self.conv_1x1_relu(c1x1) # Commented out

        c3x3 = self.conv_3x3_reduce(x)
        c3x3 = self.conv_3x3_reduce_relu(c3x3)
        c3x3 = self.conv_3x3(c3x3)
        # c3x3 = self.conv_3x3_relu(c3x3) # Commented out

        c5x5 = self.conv_5x5_reduce(x)
        c5x5 = self.conv_5x5_reduce_relu(c5x5)
        c5x5 = self.conv_5x5(c5x5)
        # c5x5 = self.conv_5x5_relu(c5x5) # Commented out

        px = self.pool(x)
        px = self.pool_proj(px)
        # px = self.pool_proj_relu(px) # Commented out
        return torch.cat([c1x1, c3x3, c5x5, px], dim=1)
```

![image](https://user-images.githubusercontent.com/13292973/112713057-0b7d3080-8f0e-11eb-93d6-7b25611fa8c7.png)

which matches the original version from the [Lucid notebook](https://colab.research.google.com/github/tensorflow/lucid/blob/master/notebooks/feature-visualization/negative_neurons.ipynb).

Reply:
@ProGamerGov  let me know if it's okay to bring the ReLU out of the InceptionModule?
I have the code ready to commit - ran our unit tests and checked through the tutorial visualizations and all looks good.

Reply:
@greentfrapp I think that the Diversity and Negative Neuron bugs had different causes. And you can bring the ReLU out the inception module!

Reply:
Just make sure that you update the `get_model_layers` test as well for the Inception5h model

Reply:
Okay just pushed a change that should fix the diversity loss. The implementation was actually a little off. The diversity loss should be based on cosine similarity between the Gram matrix representations of the activation vectors. We were using cosine similarity between activation vectors directly.
@ProGamerGov can you see if this works better?